lesson4/hometask.rb

```diff
@@ -0,0 +1,116 @@
+def test(klass)
+  lazy_hash = klass.new
+  lazy_hash[:proc] = proc { Time.now }
+  lazy_hash[:lambda] = lambda { Time.now }
+  lazy_hash[:raw] = Time.now
+
+  result = true
+
+  result = result && lazy_hash[:proc].is_a?(Proc)
+  result = result && lazy_hash[:lambda].is_a?(Proc)
+  result = result && lazy_hash[:raw].is_a?(Time)
+
+  result = result && lazy_hash.lazy?(:proc)
+  result = result && lazy_hash.lazy?(:lambda)
+  result = result && !lazy_hash.lazy?(:raw)
+
+  result = result && lazy_hash.respond_to?(:proc)
+  result = result && lazy_hash.respond_to?(:lambda)
+  result = result && lazy_hash.respond_to?(:raw)
+
+  result = result && Time.now < lazy_hash.proc
+  result = result && Time.now < lazy_hash.lambda
+  result = result && Time.now > lazy_hash.raw
+
+  result = result && lazy_hash.proc.is_a?(Time)
+  result = result && lazy_hash.lambda.is_a?(Time)
+  result = result && lazy_hash.raw.is_a?(Time)
+
+  result = result && Time.now > lazy_hash.proc
+  result = result && Time.now > lazy_hash.lambda
+  result = result && Time.now > lazy_hash.raw
+
+  result = result && lazy_hash[:proc] == lazy_hash.proc
+  result = result && lazy_hash[:lambda] == lazy_hash.lambda
+  result = result && lazy_hash[:raw] == lazy_hash.raw
+
+  result = result && !lazy_hash.lazy?(:proc)
+  result = result && !lazy_hash.lazy?(:lambda)
+  result = result && !lazy_hash.lazy?(:raw)
+
+  puts result
+end
+
+
+#==============================================
+# SimpleDelegator usage and no method missing
+#==============================================
+
+require 'delegate'
+
+class LazyHashDelegator < SimpleDelegator
+  def initialize(initial_hash = {})
+    self.__setobj__(initial_hash)
+  end
+
+  def []=(key, value)
+    raise ArgumentError, 'Unable to symbolize key!' unless key.respond_to?(:to_sym)
+    super(key.to_sym, value)
+    unless self.__getobj__.respond_to?(key.to_sym)
+      define_singleton_method(key.to_sym) do
+        obj = self.__getobj__[key]
+        obj.is_a?(Proc) ? self.__getobj__[key] = obj.call : obj
```

Comment:
There is hidden issue here.
Initially you call `super(key.to_sym, value)` that means you store `key` as symbol.
But.
In next lines you check value type. If it's a `proc` you define method to call that proc. It's ok, but how do you do this? Take a look on `self.__getobj__[key]`. Key could be a string here. And we could store value after block performing associated with string key. So, we DO NOT rewrite it, we create a new pair in that case.

Try to run this:
```ruby
lh = LazyHashDelegator.new({})
lh['test'] = proc { puts 'test output' }
p lh # => {:test=>#<Proc:0x007fd5ac04cf48@temp.rb:80>}
lh.test # => test output
p lh # => {:test=>#<Proc:0x007fd5ac04cf48@temp.rb:80>, "test"=>nil}
```